test/lit/passes/optimize-instructions-call_ref.wast

```diff
@@ -0,0 +1,201 @@
+;; NOTE: Assertions have been generated by update_lit_checks.py --all-items and should not be edited.
+;; NOTE: This test was ported using port_test.py and could be cleaned up.
+
+;; RUN: wasm-opt %s --remove-unused-names --optimize-instructions --all-features -S -o - | filecheck %s
+;; remove-unused-names is to allow fallthrough computations to work on blocks
+
+(module
+ ;; CHECK:      (type $i32_i32_=>_none (func (param i32 i32)))
+ (type $i32_i32_=>_none (func (param i32 i32)))
+
+ ;; CHECK:      (type $none_=>_i32 (func (result i32)))
+ (type $none_=>_i32 (func (result i32)))
+
+ ;; CHECK:      (type $none_=>_none (func))
+ (type $none_=>_none (func))
+
+ ;; CHECK:      (type $data_=>_none (func (param dataref)))
+ (type $data_=>_none (func (param (ref data))))
+
+ ;; CHECK:      (type $i32_=>_none (func (param i32)))
+
+ ;; CHECK:      (elem declare func $fallthrough-no-params $fallthrough-non-nullable $foo $return-nothing)
+
+ ;; CHECK:      (func $foo (param $0 i32) (param $1 i32)
+ ;; CHECK-NEXT:  (unreachable)
+ ;; CHECK-NEXT: )
+ (func $foo (param i32) (param i32)
+  (unreachable)
+ )
+ ;; CHECK:      (func $bar (param $x i32) (param $y i32)
+ ;; CHECK-NEXT:  (call $foo
+ ;; CHECK-NEXT:   (local.get $x)
+ ;; CHECK-NEXT:   (local.get $y)
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $bar (param $x i32) (param $y i32)
+  ;; This call_ref should become a direct call.
+  (call_ref
+   (local.get $x)
+   (local.get $y)
+   (ref.func $foo)
+  )
+ )
+
+ ;; CHECK:      (func $fallthrough (param $x i32)
+ ;; CHECK-NEXT:  (local $1 i32)
+ ;; CHECK-NEXT:  (call $foo
+ ;; CHECK-NEXT:   (local.tee $x
+ ;; CHECK-NEXT:    (i32.const 1)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (block (result i32)
+ ;; CHECK-NEXT:    (local.set $1
+ ;; CHECK-NEXT:     (local.get $x)
+ ;; CHECK-NEXT:    )
+ ;; CHECK-NEXT:    (drop
+ ;; CHECK-NEXT:     (block (result (ref $i32_i32_=>_none))
+ ;; CHECK-NEXT:      (local.set $x
+ ;; CHECK-NEXT:       (i32.const 2)
+ ;; CHECK-NEXT:      )
+ ;; CHECK-NEXT:      (ref.func $foo)
+ ;; CHECK-NEXT:     )
+ ;; CHECK-NEXT:    )
+ ;; CHECK-NEXT:    (local.get $1)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $fallthrough (param $x i32)
+  ;; This call_ref should become a direct call, even though it doesn't have a
+  ;; simple ref.func as the target - we need to look into the fallthrough, and
+  ;; handle things with locals.
+  (call_ref
+   ;; Write to $x before the block, and write to it in the block; we should not
+   ;; reorder these things as the side effects could alter what value appears
+   ;; in the get of $x.
```

Comment:
Is there any reordering optimization in this PR? I don't think there's one, so not sure why it is mentioned

Reply:
There isn't, but there is a risk of reordering if we naively moved the call target (the block) to before the first parameter (the tee). Instead, we append it after the second param (the get) which keeps the order. I'll clarify in the test.